### PR TITLE
New matching task tracker for task loading & tracking

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Add new matching task tracker
  * Add task iterator
  * Add task tracking functionality
  * UT

**Why?**
Part of matching logic cleanup

**How did you test it?**
UT

**Potential risks**
N/A

**Is hotfix candidate?**
No

### CODE REVIEW

**Comment by wxing1292** [comment]
Review #2262 first

File: service/matching/db_task_tracker.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		for _, task := range response.Tasks {
+			t.loadedTaskID = task.GetTaskId()
+			t.tasks[task.GetTaskId()] = false
+		}
+		if len(token) == 0 {
+			t.loadedTaskID = maxTaskID
+			if _, ok := t.tasks[maxTaskID]; !ok {
+				t.tasks[maxTaskID] = true
+			}
+		}
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/matching/db_task_tracker.go:150]: >this is important, add some comment here

sure

>It is ever possible that some new task get written later with smaller task ID than maxTaskID? I think not, just to double check.

we have a bug, if above case ever happen. the task ID min / max is derived from range ID, which is incremented each time we take ownership

File: service/matching/db_task_tracker.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	sort.Sort(taskIDs)
+
+	deleteTaskIDs := make([]int64, 0, len(taskIDs))
+	for _, taskID := range taskIDs {
+		if !t.tasks[taskID] {
+			break
+		}
+		t.ackedTaskID = taskID
+		deleteTaskIDs = append(deleteTaskIDs, taskID)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/matching/db_task_tracker.go:104]: it is usually a bad idea to delete while looping

[the remark this answers is not in the document]
Comment (yiminc) [reply service/matching/db_task_tracker.go:110]: but go explicitly allow that, in their language spec

**Review by yiminc — APPROVED** [review]

File: service/matching/db_task_tracker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/matching/db_task_tracker.go]: interesting, but you are right

**Pull request merged**